Add new findability criterion (#630)
publiccodenet/standard: criteria/findability.md

# Findability

## Requirements

* The codebase MUST be findable using a search engine by description of the problem it solves in natural language.
* The codebase MUST be findable using a search engine by codebase name.

> * Maintainers SHOULD submit the codebase to relevant software catalogs.

Reviewer: I agree, but I at the moment there's quite a bit of chaos in NL about which software catalogs exist. It would be nice if the FFPC could provide resources of which catalogs may be relevant and even actively organize these catalogs.

Reviewer 2: Yes, I agree that we should have a list of catalogs. That list might not quite fit in the catalog itself though, but perhaps a pointer to it. 

We are indeed planning a secondary resource where we provide a lot more examples of how the requirements in the standard could be met and possibly even linking to particularly elegant solutions.

Reviewer 3: I also thinking the other way around. Software catalogs should only have to know where to find repositories for indexation. If we already have the `publiccode.yaml` criterion we don't need to enlist all repositories to catalogs but insteads should be able to only submit a (organizational) github, bitbucket, gitlab etc. account from where the catalogs can start there indexation. That makes findability a organizational responsability instead of a repository maintainer one.

Reviewer 3: There is no incentive for catalog owners to change there processes when we adopt this criterium. However, when we can create an apt like structure the incentive is clear.

> * The codebase SHOULD have a website which describes the problem the codebase solves using the preferred jargon of different roles or professions

Reviewer 2: Should this perhaps end with something like "... that are users of the codebase"? This would narrow it down a bit and not suggest that all or a random set of roles and professions need to be catered for. Or is that implied enough through the context anyway?

Author: take a look now?

> * The codebase SHOULD have a website which describes the problem the codebase solves using the preferred jargon of different roles or professions

Author: ```suggestion
* The codebase SHOULD have a website which describes the problem the codebase solves using the preferred jargon of different potential users of the codebase (including technologists, policy experts and managers)
```

> * The codebase SHOULD have a website which describes the problem the codebase solves using the preferred jargon of different potential users of the codebase (including technologists, policy experts and managers.)

Author: ```suggestion
* The codebase SHOULD have a website which describes the problem the codebase solves using the preferred jargon of different potential users of the codebase (including technologists, policy experts and managers).
```

> * The codebase SHOULD have a website which describes the problem the codebase solves using the preferred jargon of different potential users of the codebase (including technologists, policy experts and managers).

Reviewer: With Open Zaak we now see that this website exists, but content-wise it's not really maintained nor clear who should maintain this. I do see the value in this, but I'm worried this will take away valuable development resources if you want to maintain this properly. So I'm not sure if this should be a SHOULD or MAY.

Reviewer 2: That is a justified worry, especially for codebase communities that haven't grown to include several active public organizations yet (or doesn't have an active product steering team). As this happens, there is a natural tendency that not everyone can contribute through developers, but that their resources at hand might be in the communications or information departments. While they might not be very technical, they will probably be able to solve everyday tasks (of course dependent on the platform used).

Reviewer 3: I don't believe this should be (solely) on a seperate website. I would like to see the information described in the repository itself. So you have one place that covers all necessary information to be able to use, implement and to asses the usefullnes of a solution.

Reviewer 2: You're correct, it shouldn't be solely on a separate website and indeed we already have requirements for this to be part of the documentation of the codebase. The documentation might be too detailed for findability and perhaps also not that easy to discover through search engines (and I don't think we necessarily want to create requirements that it should be). A landing page with a clear description with links to the repository and its documentation on its own domain or even on GH pages would be better suited for this task.

> * The codebase SHOULD have a unique and persistent identifier, such as a [Wikidata item](https://www.wikidata.org/wiki/Wikidata:Introduction) or [FSF software directory listing](https://directory.fsf.org/wiki/Main_Page), mentioning the major contributors, repository location and website.

Reviewer: I'm not opposed to this, but at first glance from both websites it's not immediately clear to me what I have to do for this. So, this requires an investment for codebases to familiarize themselves with these concepts, and the benefit is not immediately clear to me why a codebase should have this.

Reviewer 2: I agree with you here that this is a bit opaque. I think we need to add a sentence in "Why this is important" so that the value becomes clearer. Perhaps something like:

> - By having well-written metadata and being part of the semantic web, it will be easier for other people to refer, cite, disambiguate and discover the codebase through third party tools.

And I see here that we again really could use a secondary resource with good examples of this. We should perhaps raise the priority for creating it.

Reviewer 3: I would say this is a criterion for the catalog owners instead of the codebase maintainers.

> * The codebase SHOULD include a machine-readable metadata description, for example in a publiccode.yml file

Reviewer 2: ```suggestion
* The codebase SHOULD include a machine-readable metadata description, for example in a [publiccode.yml](https://github.com/publiccodeyml/publiccode.yml) file
```

Reviewer 2: I believe the link got lost in the move from the previous criteria.

> * There MAY be a dedicated domain name for the codebase.

Reviewer: How does this relate to having a website, which is a SHOULD? Wouldn't having a website imply that such a domain name exists?

Reviewer 4: @sergei-maertens consider simply having a repository well written README.md that serves the purpose, or other non-dedicated domain locations like example.org/software/foo and  example.org/software/bar ... these sites would allow a searcher to find the information, without a dedicated domain name.

Reviewer 4: For example, a project to which I contributed has a gh-pages landing page:

https://openelectronicslab.github.io/OpenHardwareExG/

Which naturally links back to the repository itself, however there is no registered domain name like OpenHardwareExG.org or similar.

* The community MAY present regularly at conferences.

## Why this is important

* Proactiveness is needed; just publishing a codebase and hoping it's found doesn't work.

> * The more findable a codebase is, the more potential new collaborators will find it.

Reviewer: Who should find the codebase? Practically in the case of Open Zaak I don't expect managers to go looking for `publiccode.yaml` on github for example. It seems that word of mouth and the VNG/GEMMA websites are the major drivers for findability.

Reviewer 2: In practice, and considering all types of codebases, I think all roles may go looking for codebases and need to find them. Yes, some codebases may be on the top-of-mind only for certain roles, and some roles are more likely to use a particular tool or method to find the codebase they need. For example, it might only be a handful of people looking for publiccode.yml file, and then it might be another, perhaps slightly larger handful, that looks in the catalogs that scrape these files whereas perhaps the majority will turn to a search engine (unless their country already have a prominent national catalog, like in Italy).

## What this does not do

* Ensure new collaborators or replicators

## How to test

Check that the codebase appears in the first set of results on more than one major search engine when you search for:

* plain problem description
* technical problem description
* codebase name

## Policy makers: what you need to do

* Contribute a description of the policy area or problem this codebase acts on or operates.
* Test your problem description with peers outside of your context who aren't familiar with the codebase.

> * Present on how the codebase implements the policy at relevent conferences.

Reviewer: Is this for the policy makers? The codebases implement policies made by policy makers, no? So I'd say this is downstream of policy makers and they are responsible for making it findable?

Reviewer 2: We meant this for policy makers, yes. When policymakers (perhaps the CIO, or even the Mayor) of a municipality developing a codebase talk about it proudly, they will probably have an audience that is more likely to be able to make decisions to join the effort. So while the downstream also is important, we wanted to address an opportunity that is often overlooked.

## Management: what you need to do

* Budget for content design and Search Engine Optimization skills in the team.
* Ensure people involved in the project present at relevant conferences.
* Search trademark databases to avoid confusion or infringement before deciding the name.

## Developers and designers: what you need to do

* Search engine optimization.
* Test your problem description with peers outside of your context who aren't familiar with the codebase.

> * Suggest conferences to present at.

Author: ```suggestion
* Suggest conferences to present at and present at them.
```

## Further reading

> * The [FAIR Guiding Principles for scientific data management and stewardship](https://www.go-fair.org/fair-principles/) provide a nice list of attributes that make (meta)data more machine actionable (and hence more findable) - some of these apply directly to codebases, while others may provoke exploration into what the codebase equivalent would be

Reviewer 2: ```suggestion
* The [FAIR Guiding Principles for scientific data management and stewardship](https://www.go-fair.org/fair-principles/) by the GO FAIR International Support and Coordination Office provide a nice list of attributes that make (meta)data more machine actionable (and hence more findable). Some of these apply directly to codebases, while others may provoke exploration into what the codebase equivalent would be.
```
